### PR TITLE
e2e improvements.

### PR DESCRIPTION
**Pull request by diegs**
Several improvements to e2e tests. Putting all in one PR to maximize probability of tests passing.

### CODE REVIEW

**Comment by diegs** [comment]
This has uncovered some other failure modes. It's very important to put update operations in retry loops in the face of transient failures.

**Comment by yifan-gu** [comment]
```
[247247.793174] golang[5]: --- FAIL: TestNetwork (92.84s)
[247247.796618] golang[5]: 	network_test.go:43: timed out waiting for nginx pod: pod not yet running: Pending
```

**Comment by yifan-gu** [comment]
Codewise LGTM.

**Comment by yifan-gu** [comment]
coreosbot run e2e

File: e2e/etcdscale_test.go [hunk 1]
```diff
@@ -3,190 +3,184 @@ package e2e
...
+	// Check for etcd-operator by getting pod
+	l, err := client.CoreV1().Pods("kube-system").List(metav1.ListOptions{LabelSelector: "k8s-app=etcd-operator"})
+	if err != nil || len(l.Items) == 0 {
+		return fmt.Errorf("test requires a cluster with self-hosted etcd: %v", err)
+	}
+
+	// Check if we are expecting enough master nodes.
 	if expectedMasters < requiredMasters {
-		t.Skip(fmt.Errorf("Test requires %d masters, test was run expecting %d", requiredMasters, expectedMasters))
+		return fmt.Errorf("Test requires %d masters, test was run expecting %d", requiredMasters, expectedMasters)
```
Comment (xiang90) [inline e2e/etcdscale_test.go]: Test -> test

File: e2e/etcdscale_test.go [hunk 2]
```diff
@@ -3,190 +3,184 @@ package e2e
...
+		}
+
+		result := httpRestClient.Put().RequestURI(kubeEtcdTPRURI).Body(data).Do()
+		if err := result.Error(); err != nil {
+			log.Printf("Error updating TPR: %v\n", err)
+			return false, nil
+		}
+		var statusCode int
+		result.StatusCode(&statusCode)
+		if statusCode != 200 {
```
Comment (xiang90) [inline e2e/etcdscale_test.go]: we need to clear the revision version before put, or there might be conflicts.

File: e2e/etcdscale_test.go (hunk 1 above)
Comment (diegs) [inline e2e/etcdscale_test.go]: Done

File: e2e/etcdscale_test.go (hunk 2 above)
Comment (xiang90) [inline e2e/etcdscale_test.go]: 200 -> http.StatusOK

File: e2e/etcdscale_test.go [hunk 3]
```diff
@@ -3,190 +3,184 @@ package e2e
...
-	// check that all pods are running by checking TPR
-	podsReady := func() error {
+	// Check that all pods are running by checking TPR.
+	if err := wait.PollImmediate(pollInterval, pollInterval, func() (bool, error) {
 		// get tpr
-		httpRestClient := c.ExtensionsV1beta1().RESTClient()
 		b, err := httpRestClient.Get().RequestURI(kubeEtcdTPRURI).DoRaw()
 		if err != nil {
-			return err
+			log.Println(err)
```
Comment (xiang90) [inline e2e/etcdscale_test.go]: add context on error logging?

Comment (diegs) [inline e2e/etcdscale_test.go]: Done

File: e2e/etcdscale_test.go (hunk 2 above)
Comment (diegs) [inline e2e/etcdscale_test.go]: Really? It should retry in that case, no?

Comment (xiang90) [inline e2e/etcdscale_test.go]: ah. ok. the get raw is inside the retry loop.

**Comment by xiang90** [comment]
lgtm

File: e2e/etcdscale_test.go (hunk 2 above)
Comment (diegs) [inline e2e/etcdscale_test.go]: Updated to `http.StatusOK`

**Pull request merged**